### PR TITLE
added cnn dataset files

### PR DESCRIPTION
**Pull request by passerim**

### CODE REVIEW

File: ellipse_detection.py
Comment (aloeser) [inline ellipse_detection.py]: just curious, where did you get that formula from?

File: ellipse_detection.py
Comment (aloeser) [inline ellipse_detection.py]: this is more @roman-staehle 's topic than mine, but I believe he used a slightly different approach do detect the coin's shapes (i.e., to find a mask for a given coin). Do you know which one works better?

File: file_loader_module.py
Comment (aloeser) [inline file_loader_module.py]: I'm not really used to chained list comprehension, but I read this as "for every path, check all the files, and for each file, do ....", i.e., i need to read it from left to right/top to bottom? Just as clarification for me because I always assumed it would be right to left

File: rotate.py
Comment (aloeser) [inline rotate.py]: Unless I overlooked something, this file is not used by any of the other files, and doesn't have a "productive" function by itself.
If we should decide to introduce the file structure that I suggested in the main review comment, how about putting this file in `src/examples` or maybe `src/experiments`?

File: scraper.py
Comment (aloeser) [inline scraper.py]: nice work, no comments or questions here :+1:

**Review by aloeser — COMMENTED** [review]
Looks good, overall! Just got a few minor comments/questions below.

But there is one general thing I'd like to discuss:
How do you guys feel about adding some folder structure to our source code? I'm not talking about some Model-View-Controller-Pattern or similar, just some structure so that we don't accumulate all our source files in the root directory.

I would suggest using the following structure, I think its pretty generic/default:

- src <- top level folder for source files, with subfolders for different tasks
  - cnn_dataset_generator
  - retrieval_dataset_generator
  - cnn

The names of the second-level-folders could are just examples to better express the idea, if you'd prefer other names we can talk about that ofc :)
The `src` folder would be on the same level as the folders containing the actual datasets

File: rotate.py
Comment (passerim) [reply rotate.py]: Yes this file was just an example. From my point of view it was intended as an example and now can be completely deleted from the repository.

File: file_loader_module.py
Comment (passerim) [reply file_loader_module.py]: You should read it as:
```
for path, _, files in os.walk(data_dir):
    for f in files:
        path.replace('\\', '/') + '/' + f
```

File: ellipse_detection.py
Comment (passerim) [reply ellipse_detection.py]: You have to provide two threshold to the Canny algorithm, a "high" primary threshold and a secondary "low" threshold, this is just a rule of thumb to pick a two reasonable values. The two values are just the upper and lower bound of the range where I expect to find 90% of the pixel intensities assuming that the intensities follow a gaussian distribution.

File: ellipse_detection.py
Comment (passerim) [reply ellipse_detection.py]: The two methods are similar, I just used some more processing steps to refine the mask using morphology. If masks obtained with the other method are not very precise, this trick could be used.

Comment (passerim) [reply ellipse_detection.py]: Maybe I'll comment something about the other method when I'll be able to review the PR for that.

**Comment by passerim** [comment]
I agree with organizing files and modules in the appropriate directories, just let me know if it's something that I should do before this PR is merged.

**Comment by roman-staehle** [comment]
I don’t know how to read/interpretate this format/answering here, but to bring some structure in general in the git repository is not a bad idea I guess 😎👍🏼

________________________________
Von: passerim ***@***.***>
Gesendet: Wednesday, June 9, 2021 2:34:45 PM
An: aloeser/CVProjectUnimore ***@***.***>
Cc: roman-staehle ***@***.***>; Mention ***@***.***>
Betreff: Re: [aloeser/CVProjectUnimore] added cnn dataset files (#3)


I agree with organizing files and modules in the appropriate directories, just let me know if it's something that I should do before this PR is merged.

File: ellipse_detection.py
Comment (aloeser) [reply ellipse_detection.py]: > when I'll be able to review the PR for that.

Should be up in the next 1-2 hours, sorry for the delay

File: ellipse_detection.py
Comment (aloeser) [reply ellipse_detection.py]: Nice, that is good to know :) Feels like i lack a lot of statistical knowledge from my home university, we had hardly any statistics courses there

File: file_loader_module.py
Comment (aloeser) [reply file_loader_module.py]: perfect,  then I understood it correctly. Thanks for the clarification!

File: rotate.py
Comment (aloeser) [reply rotate.py]: As you wish, for me both options are fine: either deleting it completely or putting it in a dedicated examples folder

**Comment by aloeser** [comment]
Nice, so we all agree on adding some folder structures!
@passerim yes, if you don't mind it would be nice if you can do that before we merge the PR, so that we dont have to touch your files afterward in an additional PR :)

**Review by aloeser — APPROVED** [review]
Nice! LGTM

File: src/cnn_dataset/scraper.py
Comment (roman-staehle) [inline src/cnn_dataset/scraper.py:62]: I checked that we said 11 countries we wanted to pick ... which was the criteria again why we chose just these countries? I kind of like the idea of adding all possible coins of the European member states 😃 - please let me know what you think about adding also the remaining countries.
Besides that we agreed on the specified range from 2002-2017 (for picture quality reasons if I remember correctly?) - This should work for all coins of all countries, the range is in general then fine with me. :-)

**Review by roman-staehle — COMMENTED** [review]
If the download of the CNN training set works for all EU countries in the specified time interval to the past, and they are outputted / saved correctly in the "data" folder (like we were able before testing on Italy/Germany) so that we can build the "retrieval data set" accordingly from them, we should be able to merge well, good work so far 👍

File: src/cnn_dataset/scraper.py
Comment (passerim) [reply src/cnn_dataset/scraper.py:62]: Adding other countries is a problem since they didn't adopt euros in 2002 but at different times, so coin images for those countries are missing or unavailable.

Comment (roman-staehle) [reply src/cnn_dataset/scraper.py:62]: alright, makes sense :-)

**Review by roman-staehle — APPROVED** [review]

**Pull request merged**
